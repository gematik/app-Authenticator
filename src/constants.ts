/*
 * Copyright 2023 gematik GmbH
 *
 * The Authenticator App is licensed under the European Union Public Licence (EUPL); every use of the Authenticator App
 * Sourcecode must be in compliance with the EUPL.
 *
 * You will find more details about the EUPL here: https://joinup.ec.europa.eu/collection/eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the EUPL is distributed on an "AS
 * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the EUPL for the specific
 * language governing permissions and limitations under the License.ee the Licence for the specific language governing
 * permissions and limitations under the Licence.
 */

export const CONFIG_FILE_NAME = 'config.json';

export const ALLOWED_DEEPLINK_PROTOCOLS = ['tim'];

export const LOG_DIRECTORY_NAME = 'authenticator-logging';

/**
 * App's slug name
 * This will be used for to open the app with parameters from the browser or any other platform.
 */
export const CUSTOM_PROTOCOL_NAME = 'authenticator';

export const CREDENTIALS_MANAGER_SERVICE_NAME = 'Gematik_Authenticator';

/**
 * Additional Information in the request header user-agent to the idp
 */
export const USERAGENT_PRODUKTNAME = 'authenticator';
export const USERAGENT_HERSTELLERNAME = 'gematik';

export const IPC_READ_MAIN_PROCESS_ENVS = 'IPC_READ_MAIN_PROCESS_ENVS';

export const IPC_UPDATE_ENV = 'IPC_UPDATE_ENV';

export const IPC_SELECT_FOLDER = 'IPC_SELECT_FOLDER';

/**
 * The event listener name for starting authentication process
 */
export const IPC_START_AUTH_FLOW_EVENT = 'IPC_START_AUTH_FLOW_EVENT';

/**
 * Event name for redirecting the http request
 */
export const IPC_GET_PATH = 'IPC_GET_PATH';

export const IPC_READ_CREDENTIALS = 'IPC_READ_CREDENTIALS';
export const IPC_SAVE_CREDENTIALS = 'IPC_SAVE_CREDENTIALS';

/**
 * This sets the user agent in the main process
 */
export const IPC_SET_USER_AGENT = 'IPC_SET_USER_AGENT';

/**
 * @deprecated remove this logic completely and use cardType parameter instead
 */
export const SCOPE_ADDITION_HBA = ' Person_ID';

/**
 * @deprecated remove this logic completely and use cardType parameter instead
 */
export const SCOPE_ADDITION_SMCB = ' Institutions_ID';

export const IPC_GET_APP_PATH = 'IPC_GET_APP_PATH';

export const IPC_GET_PROXY = 'IPC_GET_PROXY';

export const IPC_WARN_USER = 'IPC_WARN_USER';

/**
 * Start check a new update process
 */
export const IPC_CHECK_UPDATE = 'IPC_CHECK_UPDATE';
export const IPC_CANCEL_UPDATE = 'IPC_CANCEL_UPDATE';

/**
 * Event name for  maximizing and focusing to authenticator app
 */
export const IPC_FOCUS_TO_AUTHENTICATOR = 'IPC_FOCUS_TO_AUTHENTICATOR';

/**
 * Event name for  minimizing the authenticator app
 */
export const IPC_MINIMIZE_THE_AUTHENTICATOR = 'IPC_MINIMIZE_THE_AUTHENTICATOR';

/**
 * Our logging system works in the main task, not in the browser.
 * That's why we use ipc bridge and those events to send logs to file
 */
export const IPC_ERROR_LOG_EVENT_TYPES = {
  ERROR: 'IPC_ERROR_LOG_EVENT_TYPES_ERROR',
  DEBUG: 'IPC_ERROR_LOG_EVENT_TYPES_DEBUG',
  WARN: 'IPC_ERROR_LOG_EVENT_TYPES_WARN',
  INFO: 'IPC_ERROR_LOG_EVENT_TYPES_INFO',
};

/**
 * Common used regex list for javascript,
 * we use these to validate deeplink parameters and input
 */
export const COMMON_USED_REGEXES = {
  ANYTHING: /[^]*/,
  BOOLEAN_AS_TEXT: /true|false/,
  LETTERS_NUMBERS: /^[a-z0-9-_.+/=]+$/i,
  CODE: /[a-z][a-z]\w*\w\w/i,
  NUMBER: /^\d+$/,
  BASE64: /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/i,
  URL: /(http|https):\/\/(\w+:?\w*@)?(\S+)(:\d+)?(\/|\/([\w#!:.?+=&%@!\-/]))?/,
  JWT: /^[A-Za-z0-9-_=]+\.[A-Za-z0-9-_=]+\.?[A-Za-z0-9-_.+/=]*$/,
  CONNECTOR_ALLOWED: /^((?![<>&]).)*$/,
};

/**
 * Ordnername nach Setup / Name der .exe / Ordnername der Config-File mit CAs
 */
export const PRODUCT_NAME = 'gematik Authenticator';

/**
 * Name der App
 */
export const APP_NAME = 'authenticator';
/**
 * URL zu den Authenticator-Fehlercodes (Komplettübersicht)
 */
export const WIKI_ERRORCODES_URL = 'https://wiki.gematik.de/x/-A3OGw';

export const WIKI_CONSENT_DECLARATION_URL = 'https://wiki.gematik.de/x/GijaI';

/**
 * URL zum Authenticator-Sharepoint
 */
export const SHAREPOINT_URL =
  'https://gematikde.sharepoint.com/:f:/s/EXTAuthenticator/Eufh1_mbmhhImVH1ZDhgKOwBZogXt6S9PxgT7WkAY20tkA?e=1S8IR5';
/**
 * URL zur Authenticator GitHub Release Übersicht
 */
export const GITHUB_URL = 'https://github.com/gematik/app-Authenticator/releases';
/**
 * URL zum Installationshandbuch des Authenticators
 */
export const MANUAL_URL = 'https://wiki.gematik.de/x/UTlCH';
/**
 * URL zum Authenticator-Fachportal
 */
export const FACHPORTAL_URL =
  'https://fachportal.gematik.de/hersteller-anbieter/komponenten-dienste/authenticator#c6684';
/**
 * URL zum Authenticator Voraussetzungs- und Kriterienkatalog
 */
export const CRITERIA_CATALOGUE_URL = 'https://wiki.gematik.de/x/FpTRHQ';
/**
 *
 */
export const CA_CONNECTOR_DIR_PATH = 'certs-konnektor';
/**
 * Der Sub-Ordner mit CAs für IDP bei TLS-Verbindung
 */
export const CA_IDP_DIR_PATH = 'certs-idp';

export let PROCESS_ENVS: Record<string, unknown> = window?.api?.sendSync
  ? JSON.parse(window.api.sendSync(IPC_READ_MAIN_PROCESS_ENVS) as string)
  : (process.env as Record<string, unknown>);

/**
 * First re-read process envs
 */
export const updateProcessEnvs = () => {
  PROCESS_ENVS = JSON.parse(window?.api?.sendSync(IPC_READ_MAIN_PROCESS_ENVS) as string);
};

/**
 * check if env is dev
 * @deprecated IS_DEV only cen be used with MOCK_MODE == 'ENABLED' check
 */
export const IS_DEV = process.env.NODE_ENV === 'test' || PROCESS_ENVS.NODE_ENV === 'development';

/**
 * check if env is dev
 * @deprecated IS_TEST only cen be used with MOCK_MODE == 'ENABLED' check
 */
export const IS_TEST = process.env.NODE_ENV === 'test';

/**
 * Error or success message duration in the end of the auth process
 */
export const SHOW_DIALOG_DURATION = 2000;

/**
 * Retry timeout duration for auth flow
 */
export const AUTH_RE_TRY_TIMEOUT = 5000;

export const IDP_ENDPOINTS = {
  OPENID_CONFIGURATION: '/.well-known/openid-configuration',
};

export const LOGIN_VIA_SMART_CARD_SUCCESSFUL = 'login_via_smart_card_successful';

export const LOGIN_NOT_SUCCESSFUL = 'login_not_successful';

export const LOGIN_CANCELLED_BY_USER = 'login_cancelled_by_user';

export const DEV_CON_CA_CERT_PATH = '/src/assets/certs-konnektor/ca/pu/rsa';
export const DEV_IDP_CA_CERT_PATH = '/src/assets/certs-idp';

export enum P12_VALIDITY_TYPE {
  'VALID',
  'INVALID_CERTIFICATE',
  'WRONG_PASSWORD',
  'NO_CERT_FOUND',
  'ONE_VALID_AND_INVALID_CERTIFICATES',
  'TOO_MANY_CERTIFICATES',
  'PROCESSING_EXCEPTION',
}

export const STORAGE_CONFIG_KEYS = {
  HBA_CARD_USER_ID: 'userIdForCard-',
  SAVED_USER_CONSENT_PAIRS: 'savedUserConsentPairs',
  MIGRATIONS: {
    CREDENTIAL_MANAGER_STANDALONE: 'migrateCredentialManagerForStandalone',
  },
};

export const MACOS_PATHS = {
  LOGGING_DIR: '/Library/Logs/', // we add user's home path to end of this path e.x. /Users/x/Library...
  CERTS_DIR: '/Library/Application Support/' + PRODUCT_NAME,
};

export const MACOS_DS_STORE_FILE_NAME = '.DS_Store';
