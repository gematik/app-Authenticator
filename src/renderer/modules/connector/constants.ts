/*
 * Copyright 2025, gematik GmbH
 *
 * Licensed under the EUPL, Version 1.2 or - as soon they will be approved by the
 * European Commission â€“ subsequent versions of the EUPL (the "Licence").
 * You may not use this work except in compliance with the Licence.
 *
 * You find a copy of the Licence in the "Licence" file or at
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either expressed or implied.
 * In case of changes by gematik find details in the "Readme" file.
 *
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 *
 * ******
 *
 * For additional notes and disclaimer from gematik and in case of changes by gematik find details in the "Readme" file.
 */

export const CONNECTOR_SDS_PATH = '/connector.sds';

/**
 * Standard Cart Pin Types
 */
export const CARD_PIN_TYPES = {
  HBA_PIN_CH: 'PIN.CH',
  SMCB_PIN_SM: 'PIN.SMC',
};

export const XML_ERROR_TAG_NAMES = {
  TAG_ERROR_SEVERITY: 'Severity',
  TAG_ERROR_CODE: 'Code',
  TAG_ERROR_TYPE: 'ErrorType',
  TAG_ERROR_MESSAGE: 'ErrorText',
  TAG_ERROR_FAULT: 'Fault',
  TAG_ERROR_FAULT_STRING: 'faultstring',
  TAG_STATUS_ERROR: 'Error',
};

export const XML_TAG_NAMES = {
  //services
  TAG_AUTH_SIGNATURE_SERVICE: 'AuthSignatureService',
  TAG_CERTIFICATE_SERVICE: 'CertificateService',
  TAG_EVENT_SERVICE: 'EventService',
  TAG_CARD_SERVICE: 'CardService',
  //soap
  TAG_X509_CERTIFICATE: 'X509Certificate',
  TAG_BASE64SIGNATURE: 'Base64Signature',
  TAG_CARD_HANDLE: 'CardHandle',
  TAG_CARD_TERMINAL_ID: 'CtId',
  TAG_CARD_WORK_PLACE_ID: 'WorkplaceId',
  TAG_CARD_TERMINAL_WORKPLACES: 'WorkplaceIds',
  TAG_CARD_TERMINAL_NAME: 'Name',
  TAG_CARD_TERMINAL: 'CardTerminal',
  TAG_ITEM_CARD: 'Card',
  TAG_CARD_ICCSN: 'Iccsn',
  TAG_CARD_SLOT: 'SlotId',
  TAG_ENDPOINT_TLS: 'EndpointTLS',
  TAG_PINSTATUS: 'PinStatus',
  TAG_VERIFY_RESULT: 'Result',
  TAG_GETCARD_STATUS_RESULT: 'Result',
  // TAG_USERAGENT: 'node-soap/0.40.0',
  TAG_PRODUCT_TYPE_VERSION: 'ProductTypeVersion',
};

export const SOAP_ACTION_CONTENT_TYPE = 'text/xml;charset=UTF-8';

export const SOAP_ACTION = {
  ExternalAuthenticate: 'http://ws.gematik.de/conn/SignatureService/v7.4#ExternalAuthenticate',
  VerifyPin: 'http://ws.gematik.de/conn/CardService/v8.1#VerifyPin',
  GetPinStatus: 'http://ws.gematik.de/conn/CardService/v8.1#GetPinStatus',
  ReadCardCertificate: 'http://ws.gematik.de/conn/CertificateService/v6.0#ReadCardCertificate',
  GetCards: 'http://ws.gematik.de/conn/EventService/v7.2#GetCards',
  GetCardTerminals: 'http://ws.gematik.de/conn/EventService/v7.2#GetCardTerminals',
};

export enum SIGNATURE_TYPES {
  ECC = 'urn:bsi:tr:03111:ecdsa',
  RSA = 'urn:ietf:rfc:3447',
}

export enum CRYPT_TYPES {
  ECC = 'ECC',
  RSA = 'RSA',
}

export const ENCRYPTION_TYPES = {
  RSASSA_PSS_USING_SHA256: 'PS256',
  RSASSA_PKCS1_USING_SHA256: 'RS256',
  ENCRYPT_ALG_SHA256: 'SHA-256',

  ECC_ALG_SHA256: 'BP256R1',
};
