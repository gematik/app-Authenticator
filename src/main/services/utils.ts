/*
 * Copyright 2025, gematik GmbH
 *
 * Licensed under the EUPL, Version 1.2 or - as soon they will be approved by the
 * European Commission â€“ subsequent versions of the EUPL (the "Licence").
 * You may not use this work except in compliance with the Licence.
 *
 * You find a copy of the Licence in the "Licence" file or at
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either expressed or implied.
 * In case of changes by gematik find details in the "Readme" file.
 *
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

/**
 * @fileoverview Utility functions.
 */

import { START_ARGUMENTS_TO_PREVENT } from '@/constants';

/**
 * Returns true if the current platform is macOS.
 */
export const isMacOS = process.platform === 'darwin';

/**
 * Check if the app is running with any of the remote debugging flags
 */
export const hasAppRemoteDebuggingFlags = (): boolean => {
  // on development and mock mode ignore
  // #!if MOCK_MODE === 'ENABLED'
  if (process.env.NODE_ENV === 'development' || process.env.MOCK_MODE === 'ENABLED') {
    return false;
  }
  // #!endif

  let hasRemoteDebuggingFlags = false;
  for (const arg of process.argv) {
    // if app contains any of the flagsToRemove, quit the app
    if (START_ARGUMENTS_TO_PREVENT.some((flag) => arg.includes(flag))) {
      hasRemoteDebuggingFlags = true;
      break;
    }
  }

  return hasRemoteDebuggingFlags;
};
