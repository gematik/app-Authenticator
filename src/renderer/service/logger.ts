/*
 * Copyright 2024 gematik GmbH
 *
 * The Authenticator App is licensed under the European Union Public Licence (EUPL); every use of the Authenticator App
 * Sourcecode must be in compliance with the EUPL.
 *
 * You will find more details about the EUPL here: https://joinup.ec.europa.eu/collection/eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the EUPL is distributed on an "AS
 * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the EUPL for the specific
 * language governing permissions and limitations under the License.ee the Licence for the specific language governing
 * permissions and limitations under the Licence.
 */

import { IPC_ERROR_LOG_EVENT_TYPES } from '@/constants';

//TODO: @Rene: bitte im Rahmen von authcl-448 diese Function loeschen
export function getCallerInfo(err: any): string {
  const caller_line = err?.stack?.split('\n')[4];

  if (!caller_line) {
    return '';
  }

  if (Number.isNaN(caller_line)) return caller_line;
  const index = caller_line.indexOf('at ');
  return caller_line.slice(index + 2, caller_line.length);
}

export const logger = {
  error: (...args: unknown[]) => {
    // #!if MOCK_MODE === 'ENABLED'
    console.error(...args); // eslint-disable-line
    // #!endif
    window?.api?.send(IPC_ERROR_LOG_EVENT_TYPES.ERROR, args);
  },
  debug: (...args: unknown[]) => {
    // #!if MOCK_MODE === 'ENABLED'
    console.debug(...args); // eslint-disable-line
    // #!endif
    window?.api?.send(IPC_ERROR_LOG_EVENT_TYPES.DEBUG, args);
  },
  warn: (...args: unknown[]) => {
    // #!if MOCK_MODE === 'ENABLED'
    console.warn(...args); // eslint-disable-line
    // #!endif
    window?.api?.send(IPC_ERROR_LOG_EVENT_TYPES.WARN, args);
  },
  info: (...args: unknown[]) => {
    // #!if MOCK_MODE === 'ENABLED'
    console.info(...args); // eslint-disable-line
    // #!endif
    window?.api?.send(IPC_ERROR_LOG_EVENT_TYPES.INFO, args);
  },
};
