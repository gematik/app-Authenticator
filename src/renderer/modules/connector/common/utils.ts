/*
 * Copyright 2024 gematik GmbH
 *
 * The Authenticator App is licensed under the European Union Public Licence (EUPL); every use of the Authenticator App
 * Sourcecode must be in compliance with the EUPL.
 *
 * You will find more details about the EUPL here: https://joinup.ec.europa.eu/collection/eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the EUPL is distributed on an "AS
 * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the EUPL for the specific
 * language governing permissions and limitations under the License.ee the Licence for the specific language governing
 * permissions and limitations under the Licence.
 */

import parse from '@/renderer/modules/connector/common/soap-response-xml-parser';
import { XML_ERROR_TAG_NAMES } from '@/renderer/modules/connector/constants';
import { logger } from '@/renderer/service/logger';
import { readCaCerts } from '@/renderer/utils/read-tls-certificates';
import { ConnectorError, UserfacingError } from '@/renderer/errors/errors';
import { ERROR_CODES } from '@/error-codes';

export function checkSoapError(responseBody: string): ConnectorError | UserfacingError | false {
  // if there is no response body return
  if (!responseBody) {
    return false;
  }

  try {
    logger.warn(
      'Technical issue from connector. Please be sure that cards, card terminals and connector are available',
      responseBody,
    );
    if (
      responseBody.toString().includes(XML_ERROR_TAG_NAMES.TAG_ERROR_FAULT) ||
      responseBody.toString().includes(XML_ERROR_TAG_NAMES.TAG_ERROR_SEVERITY)
    ) {
      const code = parse(responseBody, XML_ERROR_TAG_NAMES.TAG_ERROR_CODE);
      const errorMessage = parse(responseBody, XML_ERROR_TAG_NAMES.TAG_ERROR_MESSAGE);
      // error = technical error
      return new ConnectorError(code, 'Technischer Fehler', errorMessage);
    }

    return new UserfacingError('Unknown Connector Error', '', ERROR_CODES.AUTHCL_1116);
  } catch (err) {
    return new UserfacingError('XML response parse error', err.message, ERROR_CODES.AUTHCL_1110);
  }
}

export function getHomedir(): string {
  return window.api.getTmpDir();
}

let CA_CHAINS_CONNECTOR: undefined | string[];

/**
 * return CA-Chain for Connector
 */
export const buildCaChainsConnector = (): undefined | string[] => {
  if (CA_CHAINS_CONNECTOR) {
    logger.debug('reuse CA_CHAINS_CONNECTOR!');
    return CA_CHAINS_CONNECTOR;
  }
  CA_CHAINS_CONNECTOR = readCaCerts(true);

  return CA_CHAINS_CONNECTOR;
};
