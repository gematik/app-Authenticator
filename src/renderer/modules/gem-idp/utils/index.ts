import { logger } from '@/renderer/service/logger';

/**
 * Send a fetch request with auto-redirects enabled
 * @param url
 */
export async function sendAutoRedirectRequest(url: string): Promise<void> {
  const response = await fetch(url);

  // Throw an error if the response status is not in the 200-399 range
  if (response.status < 200 || response.status >= 400) {
    // Customize the error message
    const error = new Error(
      `HTTP error! Status: ${response.status || 'No Status Code'}. Message: ${response.statusText || 'No Message'}`,
    );
    Object.assign(error, { response });
    logger.error(error);
    throw error;
  }
}
