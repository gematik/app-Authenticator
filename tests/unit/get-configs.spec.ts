/*
 * Copyright 2023 gematik GmbH
 *
 * The Authenticator App is licensed under the European Union Public Licence (EUPL); every use of the Authenticator App
 * Sourcecode must be in compliance with the EUPL.
 *
 * You will find more details about the EUPL here: https://joinup.ec.europa.eu/collection/eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the EUPL is distributed on an "AS
 * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the EUPL for the specific
 * language governing permissions and limitations under the License.ee the Licence for the specific language governing
 * permissions and limitations under the Licence.
 */

/**
 * @jest-environment jsdom
 */
import { setSampleData } from '../utils/config-sample-data';
import { FileStorageRepository, TRepositoryData } from '@/renderer/modules/settings/repository';
import ConnectorConfig from '@/renderer/modules/connector/connector_impl/connector-config';
import { PROCESS_ENVS, PRODUCT_NAME } from '@/constants';
import { logger } from '@/renderer/service/logger';
import { PathProvider } from '@/renderer/service/path-provider';

describe('configMapperByGroup and error handling response', () => {
  beforeEach(() => {
    // clear local storage
    const fileStorageRepository = new FileStorageRepository();
    fileStorageRepository.clear();

    setSampleData();

    // put params in connector store
    ConnectorConfig.updateConnectorParameters();
  });

  it('should set context parameters right', function () {
    expect(ConnectorConfig.contextParameters).toMatchSnapshot();
  });

  it('should set setEntryOptions right', function () {
    expect(ConnectorConfig.tlsEntryOptions).toMatchSnapshot();
  });

  it('should set auth sign params properly', function () {
    expect(ConnectorConfig.authSignParameter).toMatchSnapshot();
  });

  it('should load the config properly', function () {
    const fileStorageRepository = new FileStorageRepository();
    fileStorageRepository.clear();
    const data: TRepositoryData = fileStorageRepository.load();

    logger.info('data:' + data);
  });
});

describe('config file location test', () => {
  const env = process.env;

  beforeEach(() => {
    jest.resetModules();
    process.env = { ...env };
  });

  afterEach(() => {
    process.env = env;
  });
  const existsSyncMock = jest.fn();
  (global as any).window.api.existsSync = existsSyncMock;
  const isFile = jest.fn();
  (global as any).window.api.isFile = isFile;

  // When AUTHCONFIGPATH and COMPUTERNAME but not CLIENTNAME are set, the config-file should be located in AUTHCONFIGPATH + \\ + COMPUTERNAME
  it('test FileStorageRepository-getConfigPath-COMPUTERNAME', async () => {
    process.env.NODE_ENV = 'development';
    PROCESS_ENVS.COMPUTERNAME = 'COMPUTERNAME1';
    PROCESS_ENVS.AUTHCONFIGPATH = 'PATH1';

    existsSyncMock.mockReturnValue(true);
    isFile.mockReturnValue(true);

    const config = FileStorageRepository.getConfigPath();
    expect(config.path).toContain(PROCESS_ENVS.AUTHCONFIGPATH);
    expect(config.path).toContain(PROCESS_ENVS.COMPUTERNAME);
    expect(config.localEnv).toBeFalsy();
  });

  // When AUTHCONFIGPATH and CLIENTNAME are set, the config-file should be located in AUTHCONFIGPATH + \\ + CLIENTNAME
  it('test FileStorageRepository-getConfigPath-CLIENTNAME', async () => {
    process.env.NODE_ENV = 'development';
    PROCESS_ENVS.CLIENTNAME = 'NW001';
    PROCESS_ENVS.COMPUTERNAME = 'COMPUTERNAME1';
    PROCESS_ENVS.AUTHCONFIGPATH = 'PATH1';

    existsSyncMock.mockReturnValue(true);
    isFile.mockReturnValue(true);

    const config = FileStorageRepository.getConfigPath();

    expect(config.path).toContain(PROCESS_ENVS.AUTHCONFIGPATH);
    expect(config.path).toContain(PROCESS_ENVS.CLIENTNAME);
    expect(config.localEnv).toBeFalsy();
  });

  // When AUTHCONFIGPATH and VIEWCLIENT_MACHINE_NAME are set, the config-file should be located in AUTHCONFIGPATH + \\ + VIEWCLIENT_MACHINE_NAME
  it('test FileStorageRepository-getConfigPath-VIEWCLIENT_MACHINE_NAME', async () => {
    //process.env.NODE_ENV = 'development';
    PROCESS_ENVS.VIEWCLIENT_MACHINE_NAME = 'NW001';
    PROCESS_ENVS.COMPUTERNAME = 'COMPUTERNAME1';
    PROCESS_ENVS.AUTHCONFIGPATH = 'PATH1';

    existsSyncMock.mockReturnValue(true);
    isFile.mockReturnValue(true);

    const config = FileStorageRepository.getConfigPath();

    expect(config.path).toContain(PROCESS_ENVS.AUTHCONFIGPATH);
    expect(config.path).toContain(PROCESS_ENVS.VIEWCLIENT_MACHINE_NAME);
    expect(config.localEnv).toBeFalsy();
  });

  // When AUTHCONFIGPATH and CLIENTNAME are set and the path to the config.json did not exist then the local env has to be used
  it('test FileStorageRepository-getConfigPath-NotExists', async () => {
    process.env.NODE_ENV = 'development';
    PROCESS_ENVS.CLIENTNAME = 'NW001';
    PROCESS_ENVS.COMPUTERNAME = 'COMPUTERNAME1';
    PROCESS_ENVS.AUTHCONFIGPATH = 'PATH1';

    existsSyncMock.mockReturnValue(true);
    isFile.mockReturnValue(false);
    const sendSync = jest.fn();
    (global as any).window.api.sendSync = sendSync;
    sendSync.mockReturnValue('userData');

    const config = FileStorageRepository.getConfigPath();

    expect(config.path).toEqual('userData');
    expect(config.localEnv).toBeTruthy();
  });

  it("should return the expected config path. When TEMP-Path contains 'temp' in configPath should be only left part of string", () => {
    const expectedPath = 'C:\\Users\\username\\AppData\\Roaming\\' + PRODUCT_NAME;
    const getSystemUserTempPathSpy = jest
      .spyOn(PathProvider, 'getSystemUserTempPath')
      .mockReturnValue('C:\\Users\\username\\AppData\\Local\\Temp');
    const pathJoinSpy = jest.spyOn(window.api, 'pathJoin').mockReturnValue(expectedPath);

    const actualPath = PathProvider.configPath;

    expect(getSystemUserTempPathSpy).toHaveBeenCalled();
    expect(pathJoinSpy).toHaveBeenCalledWith('c:\\users\\username\\appdata\\local\\', PRODUCT_NAME);
    expect(actualPath).toEqual('C:\\Users\\username\\AppData\\Roaming\\' + PRODUCT_NAME);

    getSystemUserTempPathSpy.mockRestore();
    pathJoinSpy.mockRestore();
  });
});
