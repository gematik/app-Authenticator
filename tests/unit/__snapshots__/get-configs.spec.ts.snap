// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configMapperByGroup and error handling response should set auth sign params properly 1`] = `
{
  "base64data": "YK+JQHBucqT8OaqOyNHkYR4kAYtUQawphBNwfEaOA7Y=",
  "signatureCidpSchemes": "RSASSA-PSS",
  "signatureType": "urn:ietf:rfc:3447",
}
`;

exports[`configMapperByGroup and error handling response should set context parameters right 1`] = `
{
  "clientId": "Client-System-y",
  "mandantId": "Mandant-y",
  "userId": "user1",
  "workplaceId": "Arbeitsplatz-y",
}
`;

exports[`configMapperByGroup and error handling response should set setEntryOptions right 1`] = `
{
  "certFile": "c:/yyy.pem",
  "hostname": "127.1.1.1",
  "keyFile": "c:/xxx.pem",
  "localKT": "not-defined",
  "method": "GET",
  "password": "not-defined",
  "path": "/connector.sds",
  "pfxFile": "not-defined",
  "pfxPassword": "not-defined",
  "port": 80,
  "protocol": "https",
  "rejectUnauthorized": false,
  "remoteKT": "not-defined",
  "secureProtocol": "TLSv1_2_method",
  "username": "not-defined",
}
`;
